File: src/backend/src/atom.py
import json
import copy
from pathlib import Path
from typing import *


class Atom:
    def __init__(self):

        # this is the
        self.spec = {
            # general metadata
            "path": "",
            "filename": "",
            # system metadata
            "system": {
                "st_size": 0,  # integer Unix time
                "st_mtime": 0,  # integer Unix time
                "st_ctime": 0,  # integer Unix time
            },
            # shared metadata for all atoms, where possible
            "shared": {
                # general organization (e.g. Kanban board)
                "category": "",  # single string
                "tags": [],  # list of strings
                # for todo functionality
                "completed": False,  # boolean
                "deadline": 0,  # integer Unix time
            },
            "module": {
                "id": "",  # UUID string i.e. which module created it
                "name": "",  # Name of that module
                # module-specific properties below, e.g.
                # "assetsPath" : "",
                # "apiToken" : ""
            },
        }

        # this is the mutable state
        self.state = copy.deepcopy(self.spec)

    def update(self, rep: Dict) -> bool:
        # copy the original state
        save: Dict = copy.deepcopy(self.state)
        # apply the changes to the state
        save.update(rep)
        # check if the update parses correctly
        if self._parseRepresentation(save):
            self.state = save
            return True
        else:
            return False

    def load(self, path: str) -> bool:
        bundlePath: Path = Path(path).expanduser()
        # the very first thing is that the atom must be a directory, not a binary file
        if not bundlePath.is_dir():
            return False
        # the next thing is to check if the file has a manifest.json
        nextPaths = bundlePath.iterdir()
        found = False
        for p in nextPaths:
            if "manifest.json":
                found = True
                break
        if not found:
            return False
        # the next thing is to parse the manifest to JSON
        parsed: str = None
        manifestPath = bundlePath / "manifest.json"
        parsed = json.loads(manifestPath.read_text())
        if not parsed:
            return False
        # parse it to make sure it meets the spec
        if not self._parseRepresentation(parsed):
            return False
        # once all the checks are done, update state
        self.state = parsed
        return True

    def save(self, path: str) -> bool:
        # path to the bundle
        bundlePath: Path = Path(path).expanduser()
        # create the bundle if it doesnt exist
        bundlePath.mkdir(parents=True, exist_ok=True)
        # create a manifest file inside of it
        manifestPath = bundlePath / "manifest.json"
        # first check if representation is fine
        if not self._parseRepresentation(self.state):
            manifestPath.write_text(json.dumps({"error": "invalid atom"}))
            return False
        else:
            manifestPath.write_text(json.dumps(self.state))
            return True

    def toJSON(self):
        return json.dumps(self.state)

    def __str__(self):
        return str(self.toJSON())

    def _parseRepresentation(self, rep: Dict) -> bool:

        # first, see if it meets the valid spec
        for key in list(self.spec.keys()):
            firstLevelOfParsedKeys = list(rep.keys())
            if not key in firstLevelOfParsedKeys:
                return False
        # manually check the other layers
        try:
            check = rep["system"]["st_size"]
            check = rep["system"]["st_mtime"]
            check = rep["system"]["st_ctime"]
            check = rep["shared"]["category"]
            check = rep["shared"]["tags"]
            check = rep["shared"]["completed"]
            check = rep["shared"]["deadline"]
        except:
            return False
        return True

File: src/backend/atom.js
import { fs, path, Joi } from "./common";
import Spec from "./spec";

const VALID_EXTENSIONS = [".crncl", ".chaos", ".sa490"];

class Atom extends Spec {
  constructor() {
    super();
  }

  static spec = Joi.object({
    // shared metadata for all atoms, where possible
    shared: {
      // descriptors (besides the filename)
      title: Joi.string().allow(""),
      description: Joi.string().allow(""),
      image: Joi.string().allow(""),
      // general organization (e.g. Kanban board)
      category: Joi.string().allow(""),
      tags: Joi.array().items(Joi.string().allow("")),
      // for todo functionality
      completed: Joi.boolean(),
      deadline: Joi.date().timestamp(), // integer UNIX time
    },
    module: {
      id: Joi.string().required(),
      name: Joi.string().required(),
      // module specific properties:
      // assetsPath: "",
      // apiToken: ""
    },
  });

  save(pathToBundle) {
    // create the bundle dir if it doesnt exist
    if (!fs.existsSync(pathToBundle)) {
      fs.mkdirSync(pathToBundle, { recursive: true });
    }
    // create a file inside of it
    const pathToManifest = path.join(pathToBundle, "manifest.json");
    try {
      fs.writeFileSync(pathToManifest, this.toString(this.state));
    } catch (e) {
      console.log(e);
    }
  }

  load(pathToBundle) {
    // check valid atom (including manifest check)
    if (!Atom.isAtom(pathToBundle)) {
      return false;
    }
    // get the manifest
    const pathToManifest = path.join(pathToBundle, "manifest.json");
    const rawFileData = fs.readFileSync(pathToManifest);
    const jsonData = JSON.parse(rawFileData);
    // set state
    this.state = jsonData;
    return true;
  }

  static isAtom(pathToBundle) {
    // check exists
    if (!fs.existsSync(pathToBundle)) {
      return false;
    }
    // check is directory
    // the atom must be a directory, not a file
    if (fs.lstatSync(pathToBundle).isFile()) {
      return false;
    }
    // check extension
    if (!VALID_EXTENSIONS.includes(path.extname(pathToBundle))) {
      return false;
    }
    // check manifest.json
    const pathToManifest = path.join(pathToBundle, "manifest.json");
    if (!fs.existsSync(pathToManifest)) {
      return false;
    }
    // parse the representation
    const rawFileData = fs.readFileSync(pathToManifest);
    const jsonData = JSON.parse(rawFileData);
    if (!Atom.validate(jsonData)) {
      return false;
    }
    // if all checks passed, congratulations
    return true;
  }
}

export default Atom;
